Editor: Factor initial edits to own state

aduth opened the pull request:
Extracted from / complements: #10844
Partially addresses: #7409

This pull request seeks to refactor the treatment of "initial edits" out of the `state.editor.present.edits` to allow for the existing `edits` state to represent only user-initiated edits to be considered for change detection. This will enable some simplification of change detection reset conditions, and works toward #10844's goal of considering "dirtiness" as a condition of a non-empty `edits` state.

**Implementation notes:**

While I think this partly moves away from a goal of eliminating the editor setup actions, it concedes in acknowledging that there should be at most a single parse which occurs at the start of an editing session, with consideration of both edits and content. Thus, it is implemented here as part of the effect handler for `SETUP_EDITOR`. I still have some faith this could be refactored in the future, but have chosen the most direct option to supporting #10844 here.

**Testing instructions:**

Verify there are no regressions in the behavior of "initial edits", notably:

- Saving a new draft with an empty title does not show "Auto Draft" upon reload
- The demo content works as expected

Ensure tests pass:

```
npm test
```

**Follow-up tasks:**

- [ ] With improvements here to initial edits, [assignment of draft status during first auto-draft save](https://github.com/WordPress/gutenberg/blob/3600c88fda01b62463769b5e8c3e795b064e12c1/packages/editor/src/store/effects/posts.js#L69-L82) could be easily moved to the server-side setup of a new draft.
- [ ] `SETUP_EDITOR` handling of blocks should be eliminated in favor of a consistent, single `RESET_BLOCKS` action.

aduth commented:
I still need to add some test improvements here.

aduth commented:
Related: #10827

I'd incorporated `getMutateSafeObject` from there to the implementation here. I think also the `diff` and `merge` functions there will help further eliminate the redundancy between `edits` and `initialEdits` reducers.

youknowriad commented on packages/editor/src/store/reducer.js:
```diff
@@ -97,6 +101,23 @@ function getFlattenedBlocks( blocks ) {
+ *
+ * @return {Object} Mutation-safe object.
+ */
+function getMutateSafeObject( original, working ) {
```
Can we consider https://github.com/mweststrate/immer instead of this?

aduth replied:
I'm not too opposed to the idea of it as a _general_ pattern, but for one-off use it seems maybe a bit much both in terms of bundle size impact and developer knowledge barrier. Should we discuss separate to this pull request? I do think Immer seems to tackle well a pattern which is very prevalent in our stores.

youknowriad replied:
`getMutateSafeObject` can also be considered "something to learn" but yes definitely a topic for a separate discussion.

aduth replied:
There are degrees of learning 😆

youknowriad commented on packages/editor/src/store/reducer.js:
```diff
@@ -515,6 +523,51 @@ export const editor = flow( [
+
+			return action.edits;
+
+		case 'SETUP_EDITOR_STATE':
```
This is a bit confusing? Why do we need to consider two "setup" actions here? can't we just merge both in one? It probably means only use `SETUP_EDITOR_STATE` here.

aduth replied:
> This is a bit confusing? Why do we need to consider two "setup" actions here? can't we just merge both in one? It probably means only use `SETUP_EDITOR_STATE` here.

Right, this confused me as well, and I left it as a note in the original comment to circle back to this:

>Follow-up tasks:
>
>_[...]_
>
>`SETUP_EDITOR` handling of blocks should be eliminated in favor of a consistent, single `RESET_BLOCKS` action.

This is a bit of consistency for consistency's sake. You can see similar use in untouched existing reducers of `editor`.

youknowriad commented on packages/editor/src/store/reducer.js:
```diff
@@ -515,6 +523,51 @@ export const editor = flow( [
+
+			return state;
+
+		case 'UPDATE_POST':
```
Not sure I understand why we need to address this here? my thinking is that initialEdits should be "overriden" by regular edits anyway when saving, why do we need to change the "initial edits" when we "edit" a post?

aduth replied:
The issue we need to avoid is having an initial edit resurface when an `edits` key is unset by `UPDATE_POST`.

An alternative here (perhaps better even) is to unset keys from `initialEdits` when an (assumed-user-initiated) `EDIT_POST` occurs (i.e. the initial edit can be discarded at the point where a user defines their own value).

youknowriad replied:
How does a user `unset` an edit? `editPost( { something: undefined } )`? Can this be considered as revert to initial value, which means use initialEdit if defined?

aduth replied:
It becomes unset when the edit matches the saved value of the post. So if there's an initial edit lingering, it will become (wrongly) assumed as an edit.

youknowriad replied:
> It becomes unset when the edit matches the saved value of the post

Should this be: "become unset if it matches the initial edit if defined, the saved value otherwise?

Isn't it a bug now?

youknowriad replied:
(I'm fine with the current implementation, just trying to push the reasoning to see if we can't find something interesting :) )

youknowriad replied:
So say we have

```js
const savedValue = "";

const initialEdit = "Demo";
```

and if I understand properly we're going to assume we don't have anything to save if `edits` are empty.

So if the user do `edit( { title: "" } )` We shouldn't unset the edit right? because the empty title needs to be saved.

aduth replied:
> So if the user do `edit( { title: "" } )` We shouldn't unset the edit right? because the empty title needs to be saved.

This is true, and a good reason not to do the proposed refactoring of applying the difference of keys from an edit in unsetting from `initialEdits`.

But, I think in the branch as-is, this works as expected:

Initially:

```js
post = { title: '' };
initialEdits = { title: 'demo' };
edits = {};
```

Edit title to "non-empty":

```js
post = { title: '' };
initialEdits = { title: 'demo' };
edits = { title: 'non-empty' };
```

Edit title back to empty:

```js
post = { title: '' };
initialEdits = { title: 'demo' };
edits = {}; // title omitted because matches post value
```

Save will include `{ title: 'demo' }`

The `UPDATE_POST` will cause:

```js
post = { title: 'demo' }
edits = {}
initialEdits = {}
```

youknowriad replied:
> Edit title back to empty:

In this current branch at this ⬆️ point, the `isDirty` is true which means we can trigger a save but when you change the `isDirty` check to the presence of `edits`, it will be `false` right?

aduth replied:
That's correct.

youknowriad replied:
But that's wrong right? Because we still need to save the fact that we removed the "demo" from title

aduth replied:
Have I said how much I loathe our consideration of initial edits? 😄 I'll have to think on this one.

youknowriad replied:
> Is it true / should it be true that the button is enabled / disabled by whether the post is dirty? Or purely that it's saveable?

I think if the user has an empty title and a content, it should be enabled. At least that's how it works now.

aduth replied:
It becomes more an issue when considering the demo post, which should be immediately saveable even if technically `editor.edits` would be empty (since it's all initial edits).

youknowriad replied:
One thing I noticed which is weird though in our current behavior is that I'd have expected the "save" button to be disabled when you open the demo page without any modification. And if the current behavior is the right one, that I'm completely confused :P

youknowriad replied:
(I wrote the last comment without reading your response :P )

youknowriad replied:
👍

aduth commented:
I'm working to fix up the failing tests. Must have gone wrong in the rebase.

aduth committed:
Editor: Factor initial edits to own state

aduth committed:
Testing: Verify immediate saveability of demo post

youknowriad reviewed (approved):
I still don't like that the initialEdits should consider any user edits. I think we should just keep them as initial edits until the first save.

But that said, that's not a big concern, I'm approving the PR. Let's move forward.

aduth commented:
>I still don't like that the initialEdits should consider any user edits. I think we should just keep them as initial edits until the first save.

I'm not too opposed to dropping this. I don't think it has much of an impact in either case, given that user edits will always take precedence and never be unset until the save.

I'm going to do a final pass over the changes here to see if it makes sense / if we need more test coverage / etc.

aduth commented:
>I still don't like that the initialEdits should consider any user edits. I think we should just keep them as initial edits until the first save.

In looking over the proposed changes, `initialEdits` doesn't respond to edits. There should be effectively no changes to the state value between the initial setup and the first save. There might be some confusion in the `action.edits` value of the `UPDATE_POST` action, but this in-fact the save action.

aduth committed:
URL: Test undefined values as ignored

aduth committed:
URL: Test space encoding

aduth committed:
Testing: Verify saveability, non-dirtiness of initial edits

aduth commented:
I've rebased and added a number of new tests, some only incidentally related:

- 3cb31b6995fce5864499df675800f655e9aedd7a
- ddf6f115e2f49cded535080590325f396deb72cd
- 50b5c183d04e86a78117a38c1365573d58ce053b
- e024015d8b1b0759d21564dc3102b7adb19defc3

Our end-to-end utility treatment of querystrings is a bit contrary to expectations, but I opted to not mess with it too much here.
